### PR TITLE
Safe global ptrs

### LINKED ISSUE #410: ePBFT to handle the global pointers for replica, client and node safely
ePBFT sets raw global pointers for replica, node and client. We should look into either removing them from being global pointers or at least making them safe.

### PR DESCRIPTION
**Pull request by olgavrou**
Resolves #410 

- Removed `node` global pointer
- made `replica` a global unique ptr
- made all `Replica` timers unique ptrs
- replacing reference to raw pointer `replica` with `get_replica()` and `node` with `get_node()`
- using `set_node()` to convert either `replica` ptr or `client` ptr to a `node` ptr to be used where needed (client can be removed when we remove the standalone e2e pbft tests)
- when `Node` was being created it was setting the global node ptr as: `node = this`. By removing this there was a need to add extra constructors for some objects (mostly for when `Replica` is constructed) so that they take arguments for `f` and `num_of_replicas`, etc instead of doing `node->f()`,  `node->num_of_replicas()`, etc

### CODE REVIEW

File: src/consensus/pbft/libbyz/globalstate.cpp [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+#include "Client.h"
+#include "Node.h"
+#include "Replica.h"
+
+// Pointer to global replica object.
+std::unique_ptr<Replica> replica;
+std::unique_ptr<Client> client;
+Node* n;
+
+Replica* get_replica()
```
Comment (ashamis) [inline src/consensus/pbft/libbyz/globalstate.cpp]: please return a reference, this value cannot but null

File: src/consensus/pbft/libbyz/globalstate.h [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+class Replica;
+class Client;
+
+// Pointer to global replica object.
+extern std::unique_ptr<Replica> replica;
+extern std::unique_ptr<Client> client;
+
+extern Replica* get_replica();
+extern void set_node(Node* node);
+extern Node* get_node();
```
[the remark this answers is not in the document]
Comment (olgavrou) [reply src/consensus/pbft/libbyz/globalstate.h]: Thanks, taking a look now

**Comment by cimetrics** [comment]
safe_global_ptrs@[3636 aka 20200106.2](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3636&view=results) vs master ewma over 20 builds from [3546](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3546&view=results) to [3634](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3634&view=results)
images

**Comment by cimetrics** [comment]
safe_global_ptrs@[3638 aka 20200106.3](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3638&view=results) vs master ewma over 20 builds from [3546](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3546&view=results) to [3634](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3634&view=results)
images

**Pull request merged**